internal/edittree/edittree.go

```diff
@@ -0,0 +1,1161 @@
+// for an Array's child maps, where indexes that occur above the affected
+// index of the insertion/deletion must be incremented or decremented
+// appropriately. This ensures that when rendered later, the
+// original/inserted values will be spliced in at the correct offsets in
+// the final Array value.
+//
+// Due to optimizations discussed later, Array EditTree nodes do not use
+// the child key map (leaving it uninitialized), but will initialize and
+// use the child value maps normally. Array EditTree nodes are the only
+// types of EditTree nodes that should ever be expected to have initialized
+// bit-vectors present.
+//
+// # Scalar operations
+//
+// Scalars are fairly simple: just a term stored in an EditTree node, or in
+// the scalar child map of a composite type's EditTree node. They cannot
+// have children, and normally do not exist as independent EditTree nodes,
+// except to satisfy certain EditTree APIs.
+//
+// Scalar EditTree nodes can only be expected to have a valid Term value;
+// all other fields will be left uninitialized.
+//
+// # Optimization: Direct Array Indexing with ints
+//
+// Arrays are unique in Rego, because the only valid Terms that can index
+// into them are integer, numeric values. When processing the key Terms for
+// Objects and Sets, we have to identify children by their hash values
+// (which hash to integers). Because the only valid key Terms for Arrays
+// work as ints as well, we can skip the hashing step entirely, and just
+// use the int indexes *directly*.
+//
+// This provides a substantial CPU savings in benchmarks, because the
+// "index rewriting" passes become much cheaper from not having to rehash
+// every child's index.
+package edittree
+
+import (
+	"encoding/json"
+	"fmt"
+	"math/big"
+	"sort"
+	"strings"
+
+	"github.com/open-policy-agent/opa/ast"
+	"github.com/open-policy-agent/opa/internal/edittree/bitvector"
+)
+
+// Deletions are encoded with a nil value pointer.
+type EditTree struct {
+	value                *ast.Term
+	childKeys            map[int]*ast.Term
+	childScalarValues    map[int]*ast.Term
+	childCompositeValues map[int]*EditTree
+	eliminated           *bitvector.BitVector // Which original indexes have been knocked out?
+	insertions           *bitvector.BitVector // Which indexes have a new value inserted at them? (also used for "live" bookkeeping)
+}
+
+// Creates a new EditTree node from term.
+func NewEditTree(term *ast.Term) *EditTree {
+	if term == nil {
+		return nil
+	}
+
+	var tree EditTree
+	switch x := term.Value.(type) {
+	case ast.Object, ast.Set:
+		tree = EditTree{
+			value:                term,
+			childKeys:            map[int]*ast.Term{},
+			childScalarValues:    map[int]*ast.Term{},
+			childCompositeValues: map[int]*EditTree{},
+		}
+	case *ast.Array:
+		tree = EditTree{
+			value:                term,
+			childScalarValues:    map[int]*ast.Term{},
+			childCompositeValues: map[int]*EditTree{},
+		}
+		length := ((x.Len() - 1) / 8) + 1
```

Comment:
Nit: Can you use a constant that describes the `8` here.

Reply:
It's definitely an implementation detail, and was not chosen at random. I'll make a constant for it.

Reply:
I looked at the usage of the `length` variable, and chose to rename it to `bytesLength`, with a comment describing its usage in building the bit-vectors. I think that helps clarify the "why is the `8` here?" question fairly quickly for most folks just glancing at it.

@ashutosh-narkar If you think changing the `8` to a constant like `bitsPerByte` would be worth it, I'm fine doing so, I just didn't think it to be a large win on clarity over the changes mentioned above.

Reply:
That's fine.Thanks!